Move testing Parallel::mutate to ActionTesting

knelli2 opened the pull request:
## Proposed changes

~In addition to adding the function,~ this removes one of the `Parallel::mutate` functions. Previously, there were two `Parallel::mutate` functions: 1) took a reference to the GlobalCache, intended to be used in the testing framework, and 2) took a proxy to the GlobalCache, intended to be used in charm runs. However, during a charm run, there was no protection against accidentally using 1) to mutate the local cache which wouldn't broadcast the mutation to other nodes. ~Now there is only one `Parallel::mutate` function, and using the is_charm_aware function, determines if it should broadcast the mutation (for charm runs) or just mutate the local cache (testing framework).~ The `Parallel::mutate` function that was intended to be used in the testing framework was moved to the free functions of ActionTesting.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
In tests, instead of calling `Parallel::mutate(cache)`, now call `ActionTesting::mutate(cache)`.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 commented:
clang-tidy error comes from inside charm++. Is this something we are able to ignore?

kidder reviewed (commented):
changes look fine, but I'm not a big fan of the name `is_charm_aware`.  Let's discuss at tomorrow's call

knelli2 committed:
Move testing Parallel::mutate function to ActionTesting

knelli2 commented:
Discussed at the Wednesday SpECTRE meeting:

Instead of having source files deal with testing details, it was decided that the `Parallel::mutate` that was intended to be used in the testing framework should actually be moved to the testing framework. I added it to the free functions for ActionTesting because that seemed the most appropriate. Also added some docs to ActionTesting on using `ActionTesting::mutate`.

kidder reviewed (changes requested):
squash

kidder commented:
@markscheel @wthrowe @markscheel this has been updated to do what we discussed yesterday
